### ISSUE TITLE
[BUG] Large concat task time in GpuShuffleCoalesce with pinned memory pool

### ISSUE THREAD

**Issue by tgravescs** [NVIDIA/spark-rapids#1905]
**Describe the bug**
We had a case reported where the GpuShuffleCoalesce had a very large task time (51 seconds) in the concat batch phase.  This was a weird case where we had a huge number of input batches due to the number of partitions being specified but then AQE kicking in to change the number of reducers to be much less.  In this case we had like 4000ish map task and num partitions was set at 2001.  This results in the roughly 8 million columnar batches and AQE decided it only needed 167 reduce tasks instead of 2001.

```
GpuShuffleCoalesce

concat batch time total (min, med, max (stageId: taskId))
14.8 m (63 ms, 3.1 s, 51.0 s (stage 4.0: task 4336))
number of output columnar batches: 167
number of input rows: 88,213,700
number of input columnar batches: 8,211,445
```

I've only been able to reproduce it on DataProc for some reason and requires pinned memory to be enabled.
I ran some jstacks during this time and the tasks were in the pinned memory pool allocate and free.  

to reproduce:
configs:

`-executor-cores=2 --conf spark.rapids.sql.concurrentGpuTasks=2 --conf spark.sql.shuffle.partitions=2001 --conf spark.sql.adaptive.enabled=true --conf spark.rapids.memory.pinnedPool.size=2G`

```
val df = sc.makeRDD((1 to 8000000).map(x => s"bad$x"), 4102).toDF
df.filter("value != 'foo'").groupBy("value").sum().write.mode("overwrite").parquet("sumout1")
```

**Comment by tgravescs** [comment]
I was able to reproduce this locally as well, I wasn't using pinned memory locally by accident and once I figured out it was related turned it on and was able to reproduce.

**Comment by tgravescs** [comment]
The issue here is in the pinnedMemoryPool free call where it looks to combine memory sections.  This job has so many small batches that is causes that free pool to get huge and it can't combine very many at once so it ended up just iterating over and over again on this huge number of blocks:

Examples below where each iteration where it checks to combine goes over 133,000 section and each time only comes like 2. 

21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133220 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133218 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133216 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133214 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133212 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133210 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133208 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133206 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133204 iterations, cancombined was: 2
21/03/09 23:40:10 WARN PinnedMemoryPool: finished combined: 133202 iterations, cancombined was: 2

**Comment by jlowe** [comment]
Notably the data being allocated here in the pinned memory pool doesn't need to be pinned.  These are buffers coming from shuffle, and they will be concatenated on the host to another host buffer before being sent to the device.  Only the last buffer that holds the concatenated result needs to be pinned.  The other buffers are consuming pinned memory without needing the performance characteristics of pinned memory and could prevent other concurrent use-cases that benefit from pinned memory from acquiring it.

The performance of the `PinnedMemoryPool` allocation algorithm needs to be improved, but making the simple change to have CPU-side shuffle buffers not allocate pinned memory could be a more appropriate fix.

**Comment by jlowe** [comment]
Fixed by #1909
